### PR TITLE
Reimplement support for rknn detector

### PR DESCRIPTION
**Pull request by MarcA711**
This PR is not complete. I just want to get some early feedback.

One thing that bothers me is that some parts (especially the prerequisites and setup parts) of the video processing and object detection docs overlap.
Compare https://github.com/MarcA711/frigate-rockchip/blob/1e1a11ae2bffab188e788e742c9ef0f801c7d002/docs/docs/configuration/hardware_acceleration.md?plain=1#L367-L397
and https://github.com/MarcA711/frigate-rockchip/blob/1e1a11ae2bffab188e788e742c9ef0f801c7d002/docs/docs/configuration/object_detectors.md?plain=1#L317-L346

Any suggestions how to handle this better?

Generally, it would be better for the user to describe the configuration for both video processing and object detection in one place. Now, the user has to find two pages. But I think this is again rockchip specific and difficult to fix without breaking other stuff.

Moreover, was it considered to move the post-processing part of all detectors in one file? As far as I know every detector implements its own post-processing for each supported model. However, post-processing of the same model (like yolo-nas, yolox etc.) should be the same so there is some duplicate code.

### CODE REVIEW

File: frigate/detectors/plugins/rknn.py [hunk 1]
```diff
@@ -21,96 +15,146 @@ def hide_warnings(func):
...
+default_model = "default-fp16-yolonas_s"
+
+model_chache_dir = "/config/model_cache/rknn_cache/"
+
 
 class RknnDetectorConfig(BaseDetectorConfig):
     type: Literal[DETECTOR_KEY]
-    core_mask: int = Field(default=0, ge=0, le=7, title="Core mask for NPU.")
+    num_cores: int = Field(default=0, ge=0, le=3, title="Number of NPU cores to use.")
+    purge_model_cache: bool = Field(default=True)
```
Comment (NickM-27) [inline frigate/detectors/plugins/rknn.py:28]: what is this for exactly? is it to remove old RKNN model files or something that applies to new users as well?

**Comment by NickM-27** [comment]
If there are commonalities then you could put that in the installation section of the docs then have the hardware and object detector link to that as a prerequisite for specific setup

as far as post processing goes, it could make sense to have a util.py in detectors that handles this

File: frigate/detectors/plugins/rknn.py (hunk 1 above)
Comment (MarcA711) [reply frigate/detectors/plugins/rknn.py:28]: This applies to new users as well. I have to convert the models to rknn format using the same toolkit version that I use for inference. So, if I update the toolkit, the user needs to download a new updated model. This function removes older models, but the user can choose to disable this.

Comment (NickM-27) [reply frigate/detectors/plugins/rknn.py:28]: can we do something like making the name of the converted file have the version in the name so a config property is not needed and the desirable behavior is just there by default?

Comment (MarcA711) [reply frigate/detectors/plugins/rknn.py:28]: Yes, this is already the case. The naming scheme is this:
```
Filenames follow this pattern:
origin-quant-basename-soc-tk_version-rev.rknn
origin: From where comes the model? default: upstream repo; rknn: modifications from airockchip
quant: i8 or fp16
basename: e.g. yolonas_s
soc: e.g. rk3588
tk_version: e.g. v2.0.0
rev: e.g. 1
Full name could be: default-fp16-yolonas_s-rk3588-v2.0.0-1.rknn
```

However, this means that over time frigate will download more and more models and use up space.
Frigate will just purge model in the subdir `/config/model_cache/rknn_cache/". The user is not allowed to place his own models there. I will add a warning to the docs.

**Comment by MarcA711** [comment]
> as far as post processing goes, it could make sense to have a util.py in detectors that handles this

I agree. I will try to implement post-processing for all detectors. I will use the `mode: model_type:` config option that is already implemented to handle this. But I will also include a way to return the pure detections (similar to how its already done) to allow the detectors to handle post-processing differently (e.g. using opencl on GPU).

File: frigate/detectors/plugins/rknn.py (hunk 1 above)
Comment (MarcA711) [reply frigate/detectors/plugins/rknn.py:28]: Alright. Should I remove this function or set it to `false` by default?

Comment (NickM-27) [reply frigate/detectors/plugins/rknn.py:28]: I would remove it

**Comment by MarcA711** [comment]
This is now ready to merge

**Comment by blakeblackshear** [comment]
I think it's best to add some information about the license on your repo where you published the models. Also, probably best to note that here as well. We wouldn't want someone using Frigate for commercial purposes to assume that these models are free for commercial use.

**Comment by blakeblackshear** [comment]
Lastly, how painful would it be to update this so the models are downloaded at runtime? That prevents frigate from actually redistributing the models itself.

**Comment by MarcA711** [comment]
Good idea. I will add notes in my repo and frigate and I will exclude the models from the image. Maybe I will add some other models (maybe yolox) to the image later.

**Comment by MarcA711** [comment]
I added a note to my repo, the docs and to the logger on each start, so that the user knows about the license situation.

I also excluded the models from the image.

File: frigate/detectors/plugins/rknn.py [hunk 2]
```diff
@@ -1,118 +1,162 @@
...
+
+            for model_type, pattern in supported_models.items():
+                if re.match(pattern, model_path):
+                    model_matched = True
+                    model_props["model_type"] = model_type
+
+            if model_matched:
+                model_props["filename"] = model_path + f"-{soc}-v2.0.0-1.rknn"
+
+                # if model_path == default_model:
```
Comment (NickM-27) [inline frigate/detectors/plugins/rknn.py:114]: can this be removed?

Comment (MarcA711) [reply frigate/detectors/plugins/rknn.py:114]: It can be removed for now. However, I plan to include another model in the image that is freely usable. That's why I left the code there.

If it bothers you, I can remove it.

Comment (NickM-27) [reply frigate/detectors/plugins/rknn.py:114]: personally I would prefer leave it out

**Comment by MarcA711** [comment]
I see that Frigate 0.14 is coming out soon. Are there any plans for a beta phase? When should I finish my bigger changes so that they make it in that version?

**Review by NickM-27 — APPROVED** [review]

**Comment by MarcA711** [comment]
This PR is ready to merge from my side.

**Review by blakeblackshear — APPROVED** [review]

**Comment by tenten8401** [comment]
@blakeblackshear would you be open to shipping a yolo-nas model in the image if it was trained using no pre-trained weights? According to many of their github responses you aren't bound by the hostile license if you train your own model using their architecture only.

I ask because I have a super-gradients python environment set up already with an rtx 4090 and I wouldn't mind throwing a few days of compute at training a model for Frigate because it seems like a cool project :)
I'm already training a yolo-nas pose model up currently with no pre-trained weights off the coco dataset for a VR full body tracker because of the licensing.

Legally I don't think any license would apply because the super gradients python repo is a tool and the model is the output, similar to an image editor or a code compiler, but in the event that it did it'd probably be Apache 2.0: https://github.com/Deci-AI/super-gradients/issues/1993#issuecomment-2107237671

**Comment by blakeblackshear** [comment]
I would be open to it. I have a 4090 as well and plan to look at supporting yolo-nas for frigate+ too.

**Comment by tenten8401** [comment]
Sweet, I'll see if I can train up a model in a few days when mine is done :p 

What size do you think would be the best to use in Frigate? Also FYI if you're wanting to train up models with your computer as well, when you install super-gradients, you need to use @master, stable PyPi and github are both broken for various reasons.
`pip install git+https://github.com/Deci-AI/super-gradients.git@master` or clone the git repo and `pip install -e .` inside dir

**Comment by blakeblackshear** [comment]
I haven't looked at what's possible yet, but I was hoping to try a quantisized 320x320 version of the small model. Since frigate crops the image ahead of time, smaller objects aren't as important. Hard to say without comparing real world performance of a few options.

**Comment by MarcA711** [comment]
Hey @tenten8401,
this would be really great!
I don't know if you are specifically interested in yolo-nas on rockchip devices or if you just posted your idea here. I use the pre-trained image from super-gradients but I'm worried about the license. I would be interested in using your model for the rknn detector if you can provide one.

You might consider hosting a repo and upload your models + weights there. I think other people might be interested in using a pre-trained yolo-nas model without these license terms as well.

**Comment by tenten8401** [comment]
I have a 320x320 small model training now based on coco2017, but I did spot one thing in the github repo that has me a bit worried:
https://github.com/Deci-AI/super-gradients/blob/master/LICENSE.YOLONAS.md

> or any components comprising the model and the associated documentation

**Comment by MarcA711** [comment]
@tenten8401 awesome!

I am not a lawyer, but this license applies only to yolo-nas with pre-trained weights. Based on the answers to similiar questions in the super gradients repo (see here: https://github.com/Deci-AI/super-gradients/issues?q=is%3Aissue++License++Yolo-NAS), I think it is save to use your model.

Here a collaborator clearly confirms this: https://github.com/Deci-AI/super-gradients/issues/894#issuecomment-1534118271

Here is also the answer to my question from a deci employee that also says this: https://github.com/Deci-AI/super-gradients/issues/1993#issuecomment-2107237671
